include/partition/multi_level_graph.hpp

```diff
@@ -0,0 +1,191 @@
+    template <typename T> using Vector = typename util::ShM<T, UseSharedMemory>::vector;
+
+  public:
+    // We limit each node to have 255 edges
+    // this is very generous, we could probably pack this
+    using EdgeOffset = std::uint8_t;
+
+    MultiLevelGraph() = default;
+
+    MultiLevelGraph(Vector<typename SuperT::NodeArrayEntry> node_array_,
+                    Vector<typename SuperT::EdgeArrayEntry> edge_array_,
+                    Vector<EdgeOffset> node_to_edge_offset_)
+        : SuperT(std::move(node_array_), std::move(edge_array_)),
+          node_to_edge_offset(std::move(node_to_edge_offset_))
+    {
+    }
+
+    template <typename ContainerT>
+    MultiLevelGraph(const MultiLevelPartition &mlp,
+                    const std::uint32_t num_nodes,
+                    const ContainerT &edges)
+    {
+        auto highest_border_level = GetHighestBorderLevel(mlp, edges);
+        auto permutation = SortEdgesByHighestLevel(highest_border_level, edges);
+        auto sorted_edges_begin =
+            boost::make_permutation_iterator(edges.begin(), permutation.begin());
+        auto sorted_edges_end = boost::make_permutation_iterator(edges.begin(), permutation.end());
+        SuperT::InitializeFromSortedEdgeRange(num_nodes, sorted_edges_begin, sorted_edges_end);
+
+        // if the node ordering is sorting the border nodes first,
+        // the id of the maximum border node will be rather low
+        // enabling us to save some memory here
+        auto max_border_node_id = 0u;
+        for (auto edge_index : util::irange<std::size_t>(0, edges.size()))
+        {
+            if (highest_border_level[edge_index] > 0)
+            {
+                max_border_node_id =
+                    std::max(max_border_node_id,
+                             std::max(edges[edge_index].source, edges[edge_index].target));
+            }
+        }
+        BOOST_ASSERT(max_border_node_id < num_nodes);
+
+        auto edge_and_level_range = boost::combine(edges, highest_border_level);
+        auto sorted_edge_and_level_begin =
+            boost::make_permutation_iterator(edge_and_level_range.begin(), permutation.begin());
+        auto sorted_edge_and_level_end =
+            boost::make_permutation_iterator(edge_and_level_range.begin(), permutation.end());
+        InitializeOffsetsFromSortedEdges(
+            mlp, max_border_node_id, sorted_edge_and_level_begin, sorted_edge_and_level_end);
+    }
+
+    // Fast scan over all relevant border edges
+    auto GetBorderEdgeRange(const LevelID level, const NodeID node) const
+    {
+        auto begin = BeginBorderEdges(level, node);
+        auto end = SuperT::EndEdges(node);
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    // Fast scan over all relevant internal edges, that is edges that will not
+    // leave the cell of that node at the given level
+    auto GetInternalEdgeRange(const LevelID level, const NodeID node) const
```

Comment:
There is no test coverage for `GetInternalEdgeRange`.
Could you also add a test check that `GetInternalEdgeRange(level, node) ∪ GetBorderEdgeRange(level, node) = GetAdjacentEdgeRange(node)` for all nodes and all levels in the unit test graph?